File: crates/rpc/src/device_frame.rs
//! Portable DeviceRoom binary framing shared by native relays and browser clients.
//!
//! A frame is `uleb128(header_len) || UTF-8 JSON header || payload`. The
//! Worker routes using `to`/`from`; a browser client sends only `{s,k}`.

use serde::{Deserialize, Serialize};

use crate::RpcError;

/// Relay control frames have a leading space for parity with the Worker.
pub const RELAY_KIND: &str = " relay";
pub const RPC_KIND: &str = "rpc";
/// End-to-end proof that the relay can reach the remote backend host.
pub const ECHO_KIND: &str = "echo";
/// Durable Object hibernation-safe text keepalive request and response.
pub const PING_TEXT: &str = "ping";
pub const PONG_TEXT: &str = "pong";
/// The native and browser transports send both keepalive forms at this cadence.
pub const PING_INTERVAL_MS: i32 = 10_000;
/// Client-to-edge half-open transport deadline.
pub const SILENCE_LEASE_MS: i32 = 25_000;
/// Edge pongs alone do not prove the edge-to-host leg; require host echo after one succeeds.
pub const ECHO_DEADLINE_MS: i32 = 20_000;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeviceFrameHeader {
    pub s: String,
    pub k: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub to: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub from: Option<String>,
}

impl DeviceFrameHeader {
    pub fn new(s: impl Into<String>, k: impl Into<String>) -> Self {
        Self {
            s: s.into(),
            k: k.into(),
            to: None,
            from: None,
        }
    }

    pub fn with_to(mut self, conn_id: impl Into<String>) -> Self {
        self.to = Some(conn_id.into());
        self
    }
}

pub fn encode_device_frame(
    header: &DeviceFrameHeader,
    payload: &[u8],
) -> Result<Vec<u8>, RpcError> {
    let json = serde_json::to_vec(header)
        .map_err(|error| RpcError::Transport(format!("encode frame header: {error}")))?;
    let mut output = Vec::with_capacity(json.len() + payload.len() + 5);
    let mut length = json.len();
    loop {
        let mut byte = (length & 0x7f) as u8;
        length >>= 7;
        if length != 0 {
            byte |= 0x80;
        }
        output.push(byte);
        if length == 0 {
            break;
        }
    }
    output.extend_from_slice(&json);
    output.extend_from_slice(payload);
    Ok(output)
}

pub fn decode_device_frame(bytes: &[u8]) -> Result<(DeviceFrameHeader, Vec<u8>), RpcError> {
    let bad = |message: &str| RpcError::Transport(format!("device frame: {message}"));
    let mut offset = 0usize;
    let mut length = 0usize;
    let mut shift = 0u32;
    loop {
        let byte = *bytes.get(offset).ok_or_else(|| bad("truncated uleb128"))?;
        offset += 1;
        if shift >= 32 || (shift == 28 && byte & 0x70 != 0) {
            return Err(bad("uleb128 overflow"));
        }
        length |= ((byte & 0x7f) as usize) << shift;
        if byte & 0x80 == 0 {
            break;
        }
        shift += 7;
    }
    let end = offset
        .checked_add(length)
        .ok_or_else(|| bad("header length overflow"))?;
    let header: DeviceFrameHeader = serde_json::from_slice(
        bytes
            .get(offset..end)
            .ok_or_else(|| bad("truncated header"))?,
    )
    .map_err(|error| bad(&format!("bad header JSON: {error}")))?;
    Ok((header, bytes[end..].to_vec()))
}

pub fn relay_error_code(payload: &[u8]) -> Option<String> {
    #[derive(Deserialize)]
    struct RelayError {
        error: String,
    }
    serde_json::from_slice::<RelayError>(payload)
        .ok()
        .map(|error| error.error)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn browser_client_frame_has_no_routing_and_round_trips() {
        let header = DeviceFrameHeader::new(RPC_KIND, RPC_KIND);
        let bytes = encode_device_frame(&header, b"{\"id\":1}").unwrap();
        assert_eq!(bytes[0], 21); // `{"s":"rpc","k":"rpc"}`
        assert_eq!(
            decode_device_frame(&bytes).unwrap(),
            (header, b"{\"id\":1}".to_vec())
        );
    }

    #[test]
    fn golden_ts_wire_vectors_are_byte_identical() {
        let frame = encode_device_frame(&DeviceFrameHeader::new("a", RPC_KIND), &[1, 2]).unwrap();
        assert_eq!(frame, b"\x13{\"s\":\"a\",\"k\":\"rpc\"}\x01\x02");
        let routed =
            encode_device_frame(&DeviceFrameHeader::new("s1", "term").with_to("c9"), b"x").unwrap();
        assert_eq!(routed, b"\x1f{\"s\":\"s1\",\"k\":\"term\",\"to\":\"c9\"}x");
    }

    #[test]
    fn malformed_frames_fail_closed() {
        assert!(decode_device_frame(&[]).is_err());
        assert!(decode_device_frame(&[0x80, 0x80, 0x80, 0x80, 0x80]).is_err());
        assert!(decode_device_frame(&[0x95, 0x80, 0x80, 0x80, 0x10]).is_err());
        assert!(decode_device_frame(&[3, b'{', b'}']).is_err());
    }
}

File: web/packages/engine-client/src/device-frame.ts
/**
 * DeviceRoom wire framing for the browser client — a TS port of the Rust
 * codec (crates/rpc/src/device_frame.rs, shared by native relays and PR
 * #319's browser client).
 *
 * A frame is `uleb128(header_len) || UTF-8 JSON header || payload`. The
 * Worker routes on `to`/`from`; a browser client sends only `{s,k}`.
 */

import type { SocketClose, WsSocket } from "./socket";

export const RELAY_KIND = " relay";
export const RPC_KIND = "rpc";
/** End-to-end proof that the relay can reach the remote engine host. */
export const ECHO_KIND = "echo";
/** Durable Object hibernation-safe text keepalive request and response. */
export const PING_TEXT = "ping";
export const PONG_TEXT = "pong";
/** The browser transport sends both keepalive forms at this cadence. */
export const PING_INTERVAL_MS = 10_000;
/** Client-to-edge half-open transport deadline. */
export const SILENCE_LEASE_MS = 25_000;
/** Edge pongs alone do not prove the edge-to-host leg; require host echo. */
export const ECHO_DEADLINE_MS = 20_000;
/** Inbound frame ceiling (PR #319's connection.rs). */
export const MAX_FRAME = 8 * 1024 * 1024;
/** Cloudflare's per-WebSocket-message ceiling for encoded relay frames. */
export const MAX_OUTBOUND_FRAME = 1024 * 1024;
/** WebSocket high-water mark; a single relay frame may exceed it. */
export const MAX_BUFFERED = 256 * 1024;

export interface DeviceFrameHeader {
  readonly s: string;
  readonly k: string;
  readonly to?: string;
  readonly from?: string;
}

function headerBytes(header: DeviceFrameHeader): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(header));
}

/** Encode one relay frame. Throws only on a header that cannot serialize. */
export function encodeDeviceFrame(header: DeviceFrameHeader, payload: Uint8Array): Uint8Array {
  const json = headerBytes(header);
  const output = new Uint8Array(5 + json.length + payload.length);
  let length = json.length;
  let offset = 0;
  for (;;) {
    let byte = length & 0x7f;
    length >>>= 7;
    if (length !== 0) {
      byte |= 0x80;
    }
    output[offset++] = byte;
    if (length === 0) {
      break;
    }
  }
  output.set(json, offset);
  output.set(payload, offset + json.length);
  return output.slice(0, offset + json.length + payload.length);
}

/** Decode one relay frame; throws on truncated or malformed input. */
export function decodeDeviceFrame(bytes: Uint8Array): { header: DeviceFrameHeader; payload: Uint8Array } {
  let offset = 0;
  let length = 0;
  let shift = 0;
  for (;;) {
    if (offset >= bytes.length) {
      throw new Error("device frame: truncated uleb128");
    }
    const byte = bytes[offset++]!;
    if (shift >= 32 || (shift === 28 && (byte & 0x70) !== 0)) {
      throw new Error("device frame: uleb128 overflow");
    }
    length |= (byte & 0x7f) * 2 ** shift;
    if ((byte & 0x80) === 0) {
      break;
    }
    shift += 7;
  }
  const end = offset + length;
  if (end > bytes.length) {
    throw new Error("device frame: truncated header");
  }
  const headerText = new TextDecoder().decode(bytes.subarray(offset, end));
  const header = JSON.parse(headerText) as DeviceFrameHeader;
  if (typeof header?.s !== "string" || typeof header?.k !== "string") {
    throw new Error("device frame: bad header JSON");
  }
  return { header, payload: bytes.slice(end) };
}

/**
 * The browser's DeviceRoom socket: a same-origin, cookie-authenticated
 * WebSocket (`/api/browser/device/{id}/ws`) carrying RPC text inside
 * DeviceFrame binary envelopes, with ping/pong/echo liveness and
 * buffered-amount backpressure — the TS twin of PR #319's
 * `apps/web/src/rpc/connection.rs` browser transport.
 *
 * RPC text is the same protocol the direct engine listener speaks minus the
 * first-frame `Auth` envelope: the edge validated the browser session at
 * the upgrade and injected the owner identity into the DeviceRoom.
 */
export class RelaySocket implements WsSocket {
  readonly #socket: WebSocket;
  readonly #listeners = {
    open: new Set<() => void>(),
    message: new Set<(event: { data: unknown }) => void>(),
    close: new Set<(event: SocketClose) => void>(),
    error: new Set<() => void>(),
  };
  #closed = false;
  #lastRx = Date.now();
  #lastEcho = Date.now();
  readonly #heartbeat: ReturnType<typeof setInterval>;
  #drainTimer: ReturnType<typeof setInterval> | undefined;

  constructor(url: string) {
    this.#socket = new WebSocket(url);
    this.#socket.binaryType = "arraybuffer";
    this.#socket.addEventListener("open", () => {
      this.#touch();
      this.#lastEcho = Date.now();
      for (const listener of this.#listeners.open) {
        listener();
      }
    });
    this.#socket.addEventListener("message", (event: MessageEvent) => {
      if (!this.#accept(event.data)) {
        this.#protocolFailure();
      }
    });
    this.#socket.addEventListener("close", (event: CloseEvent) => {
      this.#finish();
      const close: SocketClose = {
        code: event.code,
        reason: event.reason,
        wasClean: event.wasClean,
      };
      for (const listener of this.#listeners.close) {
        listener(close);
      }
    });
    this.#socket.addEventListener("error", () => {
      for (const listener of this.#listeners.error) {
        listener();
      }
    });
    this.#heartbeat = setInterval(() => this.#tick(), PING_INTERVAL_MS);
  }

  send(data: string): void {
    const frame = encodeDeviceFrame({ s: RPC_KIND, k: RPC_KIND }, new TextEncoder().encode(data));
    if (frame.length > MAX_OUTBOUND_FRAME) {
      this.#protocolFailure();
      return;
    }
    // Backpressure mirrors the Rust client: a frame may exceed the
    // high-water mark itself; only the already-buffered amount controls
    // draining. Wait asynchronously — `send` stays synchronous per the
    // client's socket contract, frames drain in order.
    if (this.#socket.bufferedAmount > MAX_BUFFERED) {
      this.#queueFrame(frame);
      return;
    }
    this.#socket.send(frame);
  }

  close(code?: number, reason?: string): void {
    this.#finish();
    if (code === undefined) {
      this.#socket.close();
    } else {
      this.#socket.close(code, reason);
    }
  }

  addEventListener(type: "open", listener: () => void): void;
  addEventListener(type: "message", listener: (event: { data: unknown }) => void): void;
  addEventListener(type: "close", listener: (event: SocketClose) => void): void;
  addEventListener(type: "error", listener: () => void): void;
  addEventListener(type: "open" | "message" | "close" | "error", listener: never | ((event: never) => void)): void {
    this.#listeners[type as "open"].add(listener as () => void);
  }

  /** Frame acceptance: pong text, echo frames, and RPC frames only. */
  #accept(data: unknown): boolean {
    if (typeof data === "string") {
      if (data === PONG_TEXT) {
        this.#touch();
        return true;
      }
      return false;
    }
    if (!(data instanceof ArrayBuffer) || data.byteLength > MAX_FRAME) {
      return false;
    }
    let decoded: { header: DeviceFrameHeader; payload: Uint8Array };
    try {
      decoded = decodeDeviceFrame(new Uint8Array(data));
    } catch {
      return false;
    }
    if (decoded.header.k === ECHO_KIND) {
      this.#echoed();
      return true;
    }
    if (decoded.header.k === RPC_KIND) {
      this.#echoed();
      const text = new TextDecoder().decode(decoded.payload);
      for (const listener of this.#listeners.message) {
        listener({ data: text });
      }
      return true;
    }
    return false;
  }

  #tick(): void {
    if (this.#closed || this.#socket.readyState !== WebSocket.OPEN) {
      return;
    }
    const now = Date.now();
    if (now - this.#lastRx > SILENCE_LEASE_MS || now - this.#lastEcho > ECHO_DEADLINE_MS) {
      this.close(1000, "relay liveness expired");
      return;
    }
    this.#socket.send(PING_TEXT);
    this.#socket.send(encodeDeviceFrame({ s: ECHO_KIND, k: ECHO_KIND }, new Uint8Array(0)));
  }

  #queueFrame(frame: Uint8Array): void {
    if (this.#drainTimer !== undefined) {
      return;
    }
    const queue: Uint8Array[] = [frame];
    this.#drainTimer = setInterval(() => {
      if (this.#closed) {
        clearInterval(this.#drainTimer);
        this.#drainTimer = undefined;
        return;
      }
      while (this.#socket.bufferedAmount <= MAX_BUFFERED && queue.length > 0) {
        const next = queue.shift()!;
        if (next.length > MAX_OUTBOUND_FRAME) {
          this.#protocolFailure();
          return;
        }
        this.#socket.send(next);
      }
      if (queue.length === 0) {
        clearInterval(this.#drainTimer!);
        this.#drainTimer = undefined;
      }
    }, 16);
  }

  #protocolFailure(): void {
    this.close(1000, "relay protocol failure");
  }

  #touch(): void {
    this.#lastRx = Date.now();
  }

  #echoed(): void {
    this.#touch();
    this.#lastEcho = Date.now();
  }

  #finish(): void {
    if (this.#closed) {
      return;
    }
    this.#closed = true;
    clearInterval(this.#heartbeat);
    if (this.#drainTimer !== undefined) {
      clearInterval(this.#drainTimer);
      this.#drainTimer = undefined;
    }
  }
}
